modules/ref_fb_module/src/struct_decoder_fb_impl.cpp

```diff
@@ -0,0 +1,289 @@
+{
+    createInputPorts();
+    initStatuses();
+}
+
+FunctionBlockTypePtr StructDecoderFbImpl::CreateType()
+{
+    return FunctionBlockType("RefFBModuleStructDecoder", "Struct decoder", "Decodes signals with struct data type and outputs signal for each struct component.");
+}
+
+void StructDecoderFbImpl::processSignalDescriptorChanged(const DataDescriptorPtr& inputDataDescriptor,
+                                                         const DataDescriptorPtr& inputDomainDataDescriptor)
+{
+    if (inputDataDescriptor.assigned())
+        this->inputDataDescriptor = inputDataDescriptor;
+    if (inputDomainDataDescriptor.assigned())
+        this->inputDomainDataDescriptor = inputDomainDataDescriptor;
+
+    configure();
+}
+
+void StructDecoderFbImpl::configure()
+{
+    if (!inputDataDescriptor.assigned() || !inputDomainDataDescriptor.assigned())
+    {
+        setInputStatus(InputInvalid);
+        return;
+    }
+
+    try
+    {
+        if (inputDataDescriptor.getDimensions().getCount() > 0)
+            throw std::runtime_error("Arrays not supported");
+
+        const auto inputSampleType = inputDataDescriptor.getSampleType();
+        if (inputSampleType != SampleType::Struct)
+            throw std::runtime_error("Invalid sample type");
+
+        signals.clear();
+
+        constexpr std::array<SampleType, 11> validFieldTypes{
+            SampleType::Int8,
+            SampleType::Int16,
+            SampleType::Int32,
+            SampleType::Int64,
+            SampleType::UInt8,
+            SampleType::UInt16,
+            SampleType::UInt32,
+            SampleType::UInt64,
+            SampleType::Float32,
+            SampleType::Float64,
+            SampleType::Struct,
+        };
+
+        const auto domainSignal = createAndAddSignal("__domain", inputDomainDataDescriptor);
+        domainSignal.asPtr<IComponentPrivate>().unlockAttributes(List<IString>("Visible"));
+        domainSignal.setVisible(false);
+        domainSignal.asPtr<IComponentPrivate>().lockAttributes(List<IString>("Visible"));
+
+        const auto structFields = inputDataDescriptor.getStructFields();
+        for (const auto& field: structFields)
+        {
+            const auto fieldSampleType = field.getSampleType();
+            if (std::find(validFieldTypes.begin(), validFieldTypes.end(), fieldSampleType) == validFieldTypes.end())
+                throw std::runtime_error(fmt::format("Field \"{}\" has invalid sample type", field.getName()));
+
+            const auto signal = createAndAddSignal(field.getName(), field);
+            signal.setDomainSignal(domainSignal);
+        }
+
+        structSize = inputDataDescriptor.getSampleSize();
+
+        setInputStatus(InputConnected);
+    }
+    catch (const std::exception& e)
+    {
+        setInputStatus(InputInvalid);
+        LOG_W("Failed to configure output signals: {}", e.what())
+        signals.clear();
+    }
+}
+
+
+void StructDecoderFbImpl::onPacketReceived(const InputPortPtr& port)
```

Comment:
Should our examples use readers? This fb could also be written with a stream reader.

Reply:
I want to keep the same packet size/sample count on output as on input. I am not sure if this is guaranteed with stream reader. Also stream reader does not support skipping samples when input is not configured, does it?

Reply:
No, not yet. That code is in a PR at the moment.